render/pod.go

```diff
@@ -58,56 +68,58 @@ var PodServiceRenderer = ConditionalRenderer(renderKubernetesTopologies,
 		PropagateSingleMetrics(report.Pod),
 		MakeReduce(
 			MakeMap(
-				Map2Parent(report.Service, "", nil),
+				Map2Parent([]string{report.Service}, NoParentsDrop, "", nil),
 				PodRenderer,
 			),
 			SelectService,
 		),
 	),
 )
 
-// DeploymentRenderer is a Renderer which produces a renderable kubernetes deployments
-// graph by merging the pods graph and the deployments topology.
-var DeploymentRenderer = ConditionalRenderer(renderKubernetesTopologies,
-	MakeMap(
-		PropagateSingleMetrics(report.ReplicaSet),
-		MakeReduce(
-			MakeMap(
-				Map2Parent(report.Deployment, "", mapPodCounts),
-				ReplicaSetRenderer,
-			),
-			SelectDeployment,
-		),
-	),
-)
-
 // ReplicaSetRenderer is a Renderer which produces a renderable kubernetes replica sets
 // graph by merging the pods graph and the replica sets topology.
 var ReplicaSetRenderer = ConditionalRenderer(renderKubernetesTopologies,
 	MakeMap(
 		PropagateSingleMetrics(report.Pod),
 		MakeReduce(
 			MakeMap(
-				Map2Parent(report.ReplicaSet, "", nil),
+				Map2Parent([]string{report.ReplicaSet}, NoParentsDrop, "", nil),
 				PodRenderer,
 			),
 			SelectReplicaSet,
 		),
 	),
 )
 
-// DaemonSetRenderer is a Renderer which produces a renderable kubernetes daemonsets
-// graph by merging the pods graph and the daemonsets topology.
-var DaemonSetRenderer = ConditionalRenderer(renderKubernetesTopologies,
-	MakeMap(
-		PropagateSingleMetrics(report.Pod),
-		MakeReduce(
+// KubeControllerRenderer is a Renderer which combines all the 'controller' topologies.
+// We first map pods to all possible things they can map to, then we map again to
+// deployments since some things (replica sets) can map to those (the rest are passed through
+// unchanged).
+// Pods with no controller are mapped to 'Unmanaged'
+// We can't simply combine the rendered graphs of the high level objects as they would never
+// have connections to each other.
+// We combine with all the full topologies using ReduceFirstOnly, which keeps the same
+// set of nodes but merges in the full data from the other renderers.
+var KubeControllerRenderer = ConditionalRenderer(renderKubernetesTopologies,
```

Comment:
Thinking ahead to a world where probes don't include ReplicaSets, and instead list Deployments as direct parents of pods, it occurs to me that we should somehow transform the pod topology to add deployments as parents.

We can do that in the pod renderer. Or even as early as the report reading.

The rest then ought to be straightforward.